apps/nsqd/nsqd.go

```diff
@@ -246,3 +247,10 @@ func (p *program) Stop() error {
 	}
 	return nil
 }
+
+func (p *program) GetExitChan() (chan interface{}, error) {
```

Comment:
I don't think we need to change the API of `go-svc` to accomplish this. We can just pass `fatalErrCh` to `svc.Run` (it can accept multiple channels in its function signature).

Reply:
I am confused. I can't imagine how to pass `fatalErrCh` to `svc.Run` without change the API of `go-svc`,  Can you write a code example?
https://github.com/judwhite/go-svc/blob/d83e900e4a688aad49f41263bebf8793f0bf6add/svc/svc_other.go#L14-L33

Reply:
Ahhh, sorry, you're right, I misread the API.

😭

Reply:
Yeah, it takes signals, not channels.

What's wrong with `sync.Once`?

Reply:
> What's wrong with `sync.Once`?

if the nsqd `tcpServer` and `httpServer` has some error at the same time,  they will trying to close `fatalErrChan` twice, the panic will happens, so we should add sync.once make sure we will close `fatalErrChan` once

Reply:
@mreiferson @judwhite It seems that we need to change the API of `go-svc`, I have submitted my change to `https://github.com/judwhite/go-svc`, I did not expect to merge it, thanks.

Reply:
@mdh67899 Can you walk me through the problem, and proposed solution? I'm coming in with very little context, would be helpful and save me some time going through previous PR's if you can help explain the intent.

Reply:
@judwhite
In the `svc.Run()` method, after we execute the `service.Init()` and `service.Start()` method, we immediately listen to the system signal

Imagine this situation: There are some problems inside the program, we want to use `service.Stop ()` method to make sure we will  exit program more graceful

`svc.Run ()` method does not provide the notification interface of the program, we can not actively notify the service smooth exit program, we can only exit with `os.Exit ()` way

my suggestion is: we shoud  add `GetExitChan()` method inside `service`, after the program finishes `service.Init ()` and `service.Start ()`, if you receive the system signals or the exit signal given by the program, we will exit with `service.Stop ()` method

sorry for my poor english.

Reply:
@mdh67899 You could create your own channel and spin up a goroutine:

```go
go func() {
    <-quit
    service.Stop() 
}
```

Close `quit` wherever you want. Probably protect `Stop()` from executing more than once, in case that'd be a bad thing.. say if the error condition is hit at the same time a `SIGTERM` is received.

Crash-only can also be a valid strategy (`os.Exit(1)` or `log.Fatal`), though it can be a lot of effort to deliver the same experience if you don't have crash-restart-crash-restart pretty solid.

Reply:
Ahh thats a good suggestion, thanks @judwhite